core/embed/rust/src/ui/model_tt/layout.rs

```diff
@@ -1059,6 +1079,25 @@ extern "C" fn new_show_remaining_shares(n_args: usize, args: *const Obj, kwargs:
     unsafe { util::try_with_args_and_kwargs(n_args, args, kwargs, block) }
 }
 
+extern "C" fn new_show_progress(n_args: usize, args: *const Obj, kwargs: *mut Map) -> Obj {
+    let block = move |_args: &[Obj], kwargs: &Map| {
+        let title: StrBuffer = kwargs.get(Qstr::MP_QSTR_title)?.try_into()?;
+        let indeterminate: bool = kwargs.get_or(Qstr::MP_QSTR_indeterminate, false)?;
+        let description_callback: Obj = kwargs.get(Qstr::MP_QSTR_description)?;
+
+        let obj = if description_callback == Obj::const_none() {
```

Comment:
Closures are fun:)
```
error[E0308]: `if` and `else` have incompatible types
    --> src/ui/model_tt/layout.rs:1091:13
     |
1088 |            let callback = if description_callback == Obj::const_none() {
     |   ________________________-
1089 |  |             StrBuffer::empty
     |  |             ---------------- expected because of this
1090 |  |         } else {
1091 |  |             move || {
     |  |_____________^
1092 | ||                 StrBuffer::try_from(description_callback.call_with_n_args(&[]).unwrap()).unwrap()
1093 | ||             }
     | ||_____________^ expected fn item, found closure
1094 |  |         };
     |  |_________- `if` and `else` have incompatible types
     |
     = note: expected fn item `fn() -> StrBuffer {StrBuffer::empty}`
                found closure `[closure@src/ui/model_tt/layout.rs:1091:13: 1091:20]`
```

Reply:
oh great, so the two `Progress::new` calls are actually two different types.

could we declare the `description_func` as dyn somehow to avoid this? (and then probably make `move || StrBuffer::empty()` as the parameter. i mean we can probably do that in any case to make the above work, but that's not catching other uses of `Progress`)

Reply:
Because we currently don't pass any string in an event. We could add support, but then we'd have to add type parameter everywhere that touches it (or introduce dependency of ui to micropython).

Reply:
... on the other hand we don't have to store the string or do weird things with it. Perhaps we can get away by adding `&'a str` to the event variant and convert from StrBuffer in `obj_event`. This will introduce lifetime parameter but that may not be all that bad. Will try it in the afternoon.

Reply:
Ok, that seems to work, just needs to make a copy of the description when it changes: 629528f79cd326dc7e3d9d79b06d8d60fa56a730